scripts/TunnelHelpers.ps1

```diff
@@ -179,28 +180,91 @@ Function New-RandomPassword {
     )
 
-    if ($address.Contains("/")) {
-        # Trim any trailing prefix length, like "/126".
-        $address = $address.Substring(0, $address.IndexOf("/"))
+    $IPv6AddressBytes = [System.Net.IPAddress]::Parse($IPv6Address).GetAddressBytes()
+    $IPv6AddressBits = [System.BitConverter]::ToString($IPv6AddressBytes).Replace('-','')
+    $leftBits = $IPv6AddressBits.Substring(0, 16)
+    $leftBits = [System.Convert]::ToUInt64($leftBits, 16)
+    $rightBits = $IPv6AddressBits.Substring(16, 16)
+    $rightBits = [System.Convert]::ToUInt32($rightBits, 16)
+
+    # Create a bitmask for the left bits.
+    $mask = ""
+    $numLeftOneBits = [Math]::Min(64, $NumPrefixBits)
+    for ($i = 0; $i -lt $numLeftOneBits; $i++) {
+        $mask = $mask + "1"
+    }
+    $numLeftZeroBits = [Math]::Min(64 - $numLeftOneBits, 64)
+    for ($i = 0; $i -lt $numLeftZeroBits; $i++) {
+        $mask = $mask + "0"
+    }   
+    [UInt64]$leftBitmask = [System.Convert]::ToUInt64($mask, 2)
+    $maskedLeftAddress = $leftBits -band $leftBitmask
+
+    # Create a bitmask for the right bits.
+    $mask = ""
+    $numRightOneBits = [Math]::Max(0, $NumPrefixBits - 64)
+    for ($i = 0; $i -lt $numRightOneBits; $i++) {
+        $mask = $mask + "1"
+    }
+    $numRightZeroBits = [Math]::Min(64 - $numRightOneBits, 64)
+    for ($i = 0; $i -lt $numRightZeroBits; $i++) {
+        $mask = $mask + "0"
+    }   
+    [UInt64]$rightBitmask = [System.Convert]::ToUInt64($mask, 2)
+    $maskedRightAddress = $rightBits -band $rightBitmask
+
+    # Convert the two 64-bit numbers into a byte array with the bytes in the right order.
+    $prefixBytes =  New-Object Byte[] 16
+    $leftBytes = [System.BitConverter]::GetBytes($maskedLeftAddress) 
+    for ($i = 0; $i -lt 8; $i++) {
+        $prefixBytes[$i] = $leftBytes[8 - $i - 1]
     }
+    $rightBytes = [System.BitConverter]::GetBytes($maskedRightAddress) 
+    for ($i = 0; $i -lt 8; $i++) {
+        $prefixBytes[$i + 8] = $rightBytes[8 -$i - 1]
+    }
+
+    # Create an IPv6 string address from #prefixBytes.
+    $hexString = [System.BitConverter]::ToString($prefixBytes) -replace "-", ""
+    $hexString = "{0}:{1}:{2}:{3}:{4}:{5}:{6}:{7}" -f $hexString.Substring(0, 4), $hexString.Substring(4, 4), $hexString.Substring(8, 4), $hexString.Substring(12, 4), $hexString.Substring(16, 4), $hexString.Substring(20, 4), $hexString.Substring(24, 4), $hexString.Substring(28, 4)
+    $ip = [System.Net.IPAddress]::Parse($hexString)
+
+    $CIDRAddress = $ip.IPAddressToString
+    return $CIDRAddress.ToString() + '/' + $NumPrefixBits
+}
 
-    $address = $address.Replace("::", ":0:")    # Will break for addresses with multiple adjacent 0 quartets. This is unlikely for Azure addresses. 
+# Give me an IPv4 address, like 20.253.142.17 and a prefix length, like 16, and I return a valid prefix string, like "20.253.0.0/16".
+# This function isn't currently used, but it may be needed if we ever stop hardcoding the IPv4 private address.
```

Comment:
Where are we hardcoding IPv4's? I remember we do this in Unbound (DNS), but are there other places we are hard-coding IPv4's?

Reply:
I was referring to where we create the virtual network in New-Network, we specify the address prefix as "10.0.0.0/24".